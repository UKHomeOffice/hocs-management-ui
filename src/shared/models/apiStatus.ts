export default interface ApiStatus {
    display: string;
    timeoutPeriod: number;
    type: string;
}
