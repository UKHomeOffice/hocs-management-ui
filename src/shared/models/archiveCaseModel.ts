export default interface ArchiveCaseModel {
    caseReference: string;
    deleted: boolean;
}
