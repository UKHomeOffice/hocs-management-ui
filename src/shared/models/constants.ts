export const ADD_CHILD_TOPIC_ERROR_DESCRIPTION = 'Something went wrong while adding the child topic. Please try again.';
export const ADD_STANDARD_LINE_ERROR_DESCRIPTION = 'Something went wrong while adding the standard line. Please try again.';
export const ADD_TEMPLATE_ERROR_DESCRIPTION = 'Something went wrong while adding the template. Please try again.';
export const ADD_UNIT_ERROR_DESCRIPTION = 'Something went wrong while adding the unit. Please try again.';
export const ADD_USER_ERROR_DESCRIPTION = 'Something went wrong while adding the following users. Please try again.';
export const ADD_USER_ERROR_TITLE = 'There was an error adding the users';
export const DUPLICATE_CHILD_TOPIC_DESCRIPTION = 'A child topic with that name already exists';
export const DUPLICATE_UNIT_DESCRIPTION = 'A unit with those details already exists';
export const EMPTY_SUBMIT_ERROR_DESCRIPTION = 'Please select some users before submitting.';
export const EMPTY_SUBMIT_ERROR_TITLE = 'No users selected';
export const EMPTY_SUBMIT_TEAM_ERROR_DESCRIPTION = 'Please select a team before submitting.';
export const EMPTY_SUBMIT_TEAM_ERROR_TITLE = 'No team selected';
export const EMPTY_SUBMIT_DRAFT_QA_TEAM_ERROR_DESCRIPTION = 'Select a team for Initial Draft and QA response stages.';
export const EMPTY_SUBMIT_PRIVATE_MINISTER_TEAM_ERROR_DESCRIPTION = 'Select a team for Private Office and Minister sign off stages.';
export const EMPTY_SUBMIT_TOPIC_ERROR_DESCRIPTION = 'Select a topic before submitting.';
export const EMPTY_SUBMIT_TOPIC_ERROR_TITLE = 'No topic selected';
export const GENERAL_ERROR_TITLE = 'Something went wrong';
export const LOAD_PARENT_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPIC_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_CASE_TYPES_DESCRIPTION = 'There was an error retrieving the case types.  Please try refreshing the page.';
export const LOAD_TEAM_ERROR_DESCRIPTION = 'There was an error retrieving the team.  Please try refreshing the page.';
export const LOAD_TEAM_MEMBERS_ERROR_DESCRIPTION = 'There was an error retrieving the teams members.  Please try refreshing the page.';
export const LOAD_TEAMS_ERROR_DESCRIPTION = 'There was an error retrieving the teams.  Please try refreshing the page.';
export const LOAD_USERS_ERROR_DESCRIPTION = 'There was an error retrieving the list of users.  Please try refreshing the page.';
export const REMOVE_FROM_TEAM_ALLOCATED_ERROR_DESCRIPTION = 'The user cannot be removed from the team as they have cases assigned';
export const REMOVE_FROM_TEAM_ERROR_DESCRIPTION = 'There was an error removing the user from the team';
export const VALIDATION_ERROR_TITLE = "Thereâ€™s a problem";
