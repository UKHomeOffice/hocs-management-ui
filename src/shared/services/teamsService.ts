import axios from 'axios';
import Item from '../models/item';
import Team from '../models/team';

export const getTeam = (teamId: string) => new Promise<Team>((resolve, reject) =>
    axios.get(`/api/teams/${teamId}`)
        .then(response => resolve(response.data))
        .catch(error => reject(error))
);

export const getTeams = () => new Promise<[Item]>((resolve, reject) =>
    axios.get('/api/teams')
        .then(response => resolve(response.data))
        .catch(error => reject(error))
);

export const getTeamMembers = (teamId: string) => new Promise((resolve, reject) => axios
    .get(`/api/teams/${teamId}/members`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const getTeamsForUser = (userId: string) => new Promise<[Item]>((resolve, reject) => axios
    .get(`/api/teams/${userId}/teams`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const createTeam = (team: Team) => new Promise((resolve, reject) => axios
    .post(`/api/teams/unit/${team.unitUUID}`, team)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);
