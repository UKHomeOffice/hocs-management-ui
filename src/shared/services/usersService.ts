import apiClient from './apiClient';
import Item from '../models/item';

const usersApi = apiClient.createClient();

export class AddUserError extends Error {
    userToAdd: Item;
    constructor(message: string, userToAdd: Item) {
        super(message);
        this.userToAdd = userToAdd;
    }
}

export const getUsers = () => usersApi.get('/api/users');

export const addUserToTeam = (user: Item, teamId: string) => {
    return new Promise((resolve, reject) => usersApi
        .post(`/api/users/${user}/team/${teamId}`)
        .then(value => resolve(value.data))
        .catch(reason => reject(new AddUserError(reason, user))));
};

export const deleteUserFromTeam = (userId: string, teamId: string) => usersApi.delete(`/api/users/${userId}/team/${teamId}`);
