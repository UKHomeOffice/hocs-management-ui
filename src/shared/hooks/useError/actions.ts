import { FormError } from '../../models/formError';

export type AddFormError = {
    payload: FormError;
    type: 'AddFormError';
};

export type ClearError = {
    type: 'ClearError'
};

export type SetTitle = {
    payload: string;
    type: 'SetTitle';
};

export type SetDescription = {
    payload: string;
    type: 'SetDescription';
};

export type Action =
    AddFormError |
    ClearError |
    SetDescription |
    SetTitle;
