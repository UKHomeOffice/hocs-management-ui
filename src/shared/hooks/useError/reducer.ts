import { Action } from './actions';
import PageError from '../../models/pageError';

export const reducer = (state: PageError, action: Action): PageError => {
    switch (action.type) {
        case 'AddFormError':
            return {
                error: { ...state.error, formErrors: [...state.error ? state.error.formErrors || [] : [], action.payload] }
            };
        case 'ClearError':
            return { error: undefined };
        case 'SetDescription':
            return { error: { ...state.error, description: action.payload } };
        case 'SetTitle':
            return { error: { ...state.error, title: action.payload } };
    }
    return state;
};
