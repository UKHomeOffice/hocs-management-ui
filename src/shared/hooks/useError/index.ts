import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
