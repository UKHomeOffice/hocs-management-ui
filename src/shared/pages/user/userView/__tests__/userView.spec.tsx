import React from 'react';
import { match, MemoryRouter } from 'react-router-dom';
import { createBrowserHistory, History, Location } from 'history';
import { act, fireEvent, getByText, render, RenderResult, wait } from '@testing-library/react';
import UserView from '../userView';
import * as TeamsService from '../../../../services/teamsService';
import * as UsersService from '../../../../services/usersService';
import * as useError from '../../../../hooks/useError';
import { GENERAL_ERROR_TITLE, REMOVE_FROM_TEAM_ALLOCATED_ERROR_DESCRIPTION, REMOVE_FROM_TEAM_ERROR_DESCRIPTION, VALIDATION_ERROR_TITLE } from '../../../../models/constants';
let match: match<any>;
let history: History<any>;
let location: Location;

const userUUID = 'xx-xx-xx-xx';

jest.mock('../../../../services/teamsService', () => ({
    __esModule: true,
    getTeamsForUser: jest.fn().mockReturnValue(Promise.resolve(
        [{
            label: '__team1__',
            value: '__team1__'
        }, {
            label: '__team2__',
            value: '__team2__'
        }]
    ))
}));

jest.mock('../../../../services/usersService', () => ({
    __esModule: true,
    addUsersToTeam: jest.fn().mockReturnValue(Promise.resolve()),
    deleteUserFromTeam: jest.fn().mockReturnValue(Promise.resolve()),
    getUser: jest.fn().mockReturnValue(Promise.resolve(
        {
            label: '__user__',
            value: '__user__'
        }
    ))
}));

const getUserSpy = jest.spyOn(UsersService, 'getUser');
const getTeamsForUserSpy = jest.spyOn(TeamsService, 'getTeamsForUser');
const deleteUserFromTeamSpy = jest.spyOn(UsersService, 'deleteUserFromTeam');
const useErrorSpy = jest.spyOn(useError, 'default');
const clearErrorsSpy = jest.fn();
const setMessageSpy = jest.fn();
clearErrorsSpy.mockReset();
setMessageSpy.mockReset();

const renderComponent = () => render(
    <MemoryRouter>
        <UserView history={history} location={location} match={match}/>
    </MemoryRouter>
);

beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: { userId: userUUID },
        path: '',
        url: ''
    };

    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };

    useErrorSpy.mockImplementation(() => [{}, jest.fn(), clearErrorsSpy, setMessageSpy]);
    clearErrorsSpy.mockReset();
    setMessageSpy.mockReset();
});

describe('when the teamView component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(3);
        let wrapper: RenderResult;
        act(() => {
            wrapper = renderComponent();
        });

        await wait(() => {
            expect(getUserSpy).toHaveBeenCalled();
            expect(getTeamsForUserSpy).toHaveBeenCalled();
            expect(wrapper.container).toMatchSnapshot();
        });
    });
});

describe('when the Add team members button is clicked', () => {
    it('should push a new page into the history', async () => {
        history.push = jest.fn();
        let wrapper: RenderResult;
        act(() => {
            wrapper = renderComponent();
        });

        await wait(async () => {
            const addTeamMembersButton = getByText(wrapper.container, 'Add teams');
            fireEvent.click(addTeamMembersButton);
        });

        expect(history.push).toHaveBeenCalledWith('/user/' + userUUID + '/add-teams');
    });
});

describe('when the remove user button is clicked', () => {
    it('should remove the row from the users table', async () => {
        let wrapper: RenderResult;
        act(() => {
            wrapper = renderComponent();
        });

        await wait(async () => {
            const selectedUser = getByText(wrapper.container, '__team1__');
            const row = (selectedUser.closest('tr'));
            const removeButton = getByText(row as HTMLElement, 'Remove');
            fireEvent.click(removeButton);
        });

        expect(deleteUserFromTeamSpy).nthCalledWith(1, userUUID, '__team1__');
        expect(getTeamsForUserSpy).nthCalledWith(1, userUUID);

    });

    describe('and the service call fails', () => {
        beforeEach(() => {
            let wrapper: RenderResult;
            act(() => {
                wrapper = renderComponent();
            });

            wait(async () => {
                const selectedUser = getByText(wrapper.container, '__user1__');
                const row = (selectedUser.closest('tr'));
                const removeButton = getByText(row as HTMLElement, 'Remove');
                fireEvent.click(removeButton);
            });
        });

        describe('and its a 500 error', () => {
            beforeAll(() => {
                jest.spyOn(UsersService, 'deleteUserFromTeam').mockImplementation(() => Promise.reject({ response: { status: 500 } }));
            });
            it('should set the error state', () => {
                wait(async () => {
                    expect(setMessageSpy).toHaveBeenCalledWith({ description: REMOVE_FROM_TEAM_ERROR_DESCRIPTION, title: GENERAL_ERROR_TITLE });
                });
            });
            it('should call the clear errors method', () => {
                wait(async () => {
                    expect(clearErrorsSpy).toHaveBeenCalled();
                });
            });
        });
        describe('and its a 409', () => {
            beforeAll(() => {
                jest.spyOn(UsersService, 'deleteUserFromTeam').mockImplementation(() => Promise.reject({ response: { status: 409 } }));
            });

            it('should set the error state', () => {
                wait(async () => {
                    expect(setMessageSpy).toHaveBeenCalledWith({ description: REMOVE_FROM_TEAM_ALLOCATED_ERROR_DESCRIPTION, title: VALIDATION_ERROR_TITLE });
                });
            });
        });
    });
});

describe('when the Amend Details button is clicked', () => {
    it('should push a new page into the history', async () => {
        history.push = jest.fn();
        let wrapper: RenderResult;
        act(() => {
            wrapper = renderComponent();
        });

        await wait(async () => {
            const addTeamMembersButton = getByText(wrapper.container, 'Amend Details');
            fireEvent.click(addTeamMembersButton);
        });

        expect(history.push).toHaveBeenCalledWith('/user/' + userUUID + '/amend');
    });
});
