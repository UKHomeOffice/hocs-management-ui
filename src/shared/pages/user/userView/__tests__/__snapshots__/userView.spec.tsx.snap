// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the teamView component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-form-group"
  >
    <a
      class="govuk-back-link"
      href="/user-search"
    >
      Back
    </a>
    <div>
      <h1
        class="govuk-heading-xl"
      >
        Manage User
      </h1>
      <form>
        <p>
          <strong
            class="govuk-tag govuk-tag--grey"
          >
            Disabled
          </strong>
        </p>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="username"
            id="username-label"
          >
            Username
          </label>
          <input
            class="govuk-input"
            disabled=""
            id="username"
            name="username"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="email"
            id="email-label"
          >
            Email Address
          </label>
          <input
            class="govuk-input"
            disabled=""
            id="email"
            name="email"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="firstName"
            id="firstName-label"
          >
            First Name
          </label>
          <input
            class="govuk-input"
            disabled=""
            id="firstName"
            name="firstName"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="lastName"
            id="lastName-label"
          >
            Last Name
          </label>
          <input
            class="govuk-input"
            disabled=""
            id="lastName"
            name="lastName"
            type="text"
            value=""
          />
        </div>
      </form>
      <div
        class="govuk-!-padding-top-5"
      >
        <button
          class="govuk-button govuk-!-margin-right-1"
          data-module="govuk-button"
          type="submit"
        >
          Amend Details
        </button>
      </div>
      <table
        class="govuk-table"
      >
        <thead
          class="govuk-table__head"
        >
          <tr
            class="govuk-table__row"
          >
            <th
              class="govuk-table__header"
              scope="col"
            >
              Teams
            </th>
            <th
              class="govuk-table__header"
              scope="col"
            >
              Action
            </th>
          </tr>
        </thead>
        <tbody
          class="govuk-table__body"
        >
          <tr
            class="govuk-table__row"
          >
            <td
              class="govuk-table__cell"
            >
              __team1__
            </td>
            <td
              class="govuk-table__cell"
            >
              <a
                href="#"
              >
                Remove
              </a>
            </td>
          </tr>
          <tr
            class="govuk-table__row"
          >
            <td
              class="govuk-table__cell"
            >
              __team2__
            </td>
            <td
              class="govuk-table__cell"
            >
              <a
                href="#"
              >
                Remove
              </a>
            </td>
          </tr>
        </tbody>
      </table>
      <button
        class="govuk-button govuk-!-margin-right-1 add-user-members-button"
        data-module="govuk-button"
        type="submit"
      >
        Add teams
      </button>
    </div>
  </div>
</div>
`;
